src/Silverback.Integration/Messaging/Inbound/ErrorHandling/ErrorPoliciesHelper.cs

```diff
@@ -10,13 +10,32 @@ namespace Silverback.Messaging.Inbound.ErrorHandling
 {
     internal static class ErrorPoliciesHelper
     {
-        public static async Task<bool> ApplyErrorPoliciesAsync(ConsumerPipelineContext context, Exception exception)
+        public static async Task<bool> ApplyErrorPoliciesAsync(
+            ConsumerPipelineContext context,
+            Exception exception)
         {
-            var failedAttempts = context.Consumer.IncrementFailedAttempts(context.Envelope);
+            int failedAttempts = context.Consumer.IncrementFailedAttempts(context.Envelope);
 
             context.Envelope.Headers.AddOrReplace(DefaultMessageHeaders.FailedAttempts, failedAttempts);
+            context.Envelope.Headers.AddOrReplace(DefaultMessageHeaders.InvalidReason, exception.Message);
 
-            var errorPolicyImplementation = context.Envelope.Endpoint.ErrorPolicy.Build(context.ServiceProvider);
+            if (context.Envelope.Headers.Contains(DefaultMessageHeaders.TimestampKey))
+            {
+                context.Envelope.Headers.AddOrReplace(
+                    DefaultMessageHeaders.SourceTimestamp,
+                    context.Envelope.Headers[DefaultMessageHeaders.TimestampKey]);
+                context.Envelope.Headers.Remove(DefaultMessageHeaders.TimestampKey);
```

Comment:
We shouldn't modify the inbound envelope (representing the consumed message). We should only discard the header when mapping to the message to be produced.

Reply:
I moved this logic in the `KafkaBrokerHeadersEnricher` class (I hope that the class name sounds good).